### PR TITLE
Added new packages borgbackup and its dependency libacl

### PR DESCRIPTION
**Pull request by amaramrahul**

### CODE REVIEW

**Comment by Grimler91** [comment]
Hi, it looks like borgbackup is licensed under the BSD 3 clause license (though with small modifications/reformulation to 3rd point(?)), so `TERMUX_PKG_LICENSE="BSD 3-Clause"` should be set.

That being said, we don't usually package python packages. Are there any problems when trying to install borg through pip (after building/installing dependencies)?

**Comment by xeffyr** [comment]
Package uses native extensions and can't be cross compiled properly.

```
./data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/platform/linux.cpython-38-x86_64-linux-gnu.so
./data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/platform/posix.cpython-38-x86_64-linux-gnu.so
./data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/hashindex.cpython-38-x86_64-linux-gnu.so
./data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/crypto/low_level.cpython-38-x86_64-linux-gnu.so
./data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/chunker.cpython-38-x86_64-linux-gnu.so
./data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/compress.cpython-38-x86_64-linux-gnu.so
./data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/item.cpython-38-x86_64-linux-gnu.so
./data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/algorithms/checksums.cpython-38-x86_64-linux-gnu.so
./data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/algorithms/msgpack/_packer.cpython-38-x86_64-linux-gnu.so
./data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/algorithms/msgpack/_unpacker.cpython-38-x86_64-linux-gnu.so
```

File: packages/borgbackup/build.sh [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  export LDFLAGS+=" -lpython3.8"
+  export LDSHARED="$CC -shared"
+  export BORG_OPENSSL_PREFIX="/data/data/com.termux/files/usr"
+  export BORG_LIBLZ4_PREFIX="/data/data/com.termux/files/usr"
+  export BORG_LIBZSTD_PREFIX="/data/data/com.termux/files/usr"
+  python3.8 setup.py install --prefix=$TERMUX_PREFIX --force
+}
+
+termux_step_post_massage () {
+  find . -path '*/__pycache__*' -delete
```
Comment (xeffyr) [inline packages/borgbackup/build.sh]: pycache should not be deleted.

File: packages/borgbackup/build.sh [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+TERMUX_PKG_DEPENDS="libacl,liblz4,zstd,openssl,python"
+TERMUX_PKG_SRCURL=https://github.com/borgbackup/borg/releases/download/${TERMUX_PKG_VERSION}/borgbackup-${TERMUX_PKG_VERSION}.tar.gz
+TERMUX_PKG_BUILD_IN_SRC=true
+
+termux_step_make_install () {
+  export PYTHONPATH=$TERMUX_PREFIX/lib/python3.8/site-packages
+  export CPPFLAGS+=" -I$TERMUX_PREFIX/include/python3.8"
+  export LDFLAGS+=" -lpython3.8"
+  export LDSHARED="$CC -shared"
+  export BORG_OPENSSL_PREFIX="/data/data/com.termux/files/usr"
```
Comment (xeffyr) [inline packages/borgbackup/build.sh]: `/data/data/com.termux/files/usr` should be replaced as `$TERMUX_PREFIX`.

**Comment by amaramrahul** [comment]
> Hi, it looks like borgbackup is licensed under the BSD 3 clause license (though with small modifications/reformulation to 3rd point(?)), so `TERMUX_PKG_LICENSE="BSD 3-Clause"` should be set.
> 
> That being said, we don't usually package python packages. Are there any problems when trying to install borg through pip (after building/installing dependencies)?

Just gave it a shot.  Installed all the packages mentioned in scripts/setup-termux.sh and then ran:

$ env BORG_OPENSSL_PREFIX="/data/data/com.termux/files/usr" BORG_LIBLZ4_PREFIX="/data/data/com.termux/files/usr" BORG_LIBZSTD_PREFIX="/data/data/com.termux/files/usr" pip install borgbackup

borg got successfully installed. Its dependency libacl could also probably be installed using ./configure, make and make install.

However, installing all these additional packages, using the custom environment and compiling from source would be pretty difficult for the regular user.

**Comment by xeffyr** [comment]
Cross-compilation and building on device are different things.
Python does not behave well when cross-compiling package with native extensions (that is why we actually don't like adding python packages).

**Comment by xeffyr** [comment]
borg a patch to change libc.so name, otherwise it fails at runtime:
```
Exception: Can't find C library [dlopen failed: library "libc.so.6" not found]. Try installing ldconfig, gcc/cc or objdump.
```

**Comment by amaramrahul** [comment]
> borg a patch to change libc.so name, otherwise it fails at runtime:
> 
> ```
> Exception: Can't find C library [dlopen failed: library "libc.so.6" not found]. Try installing ldconfig, gcc/cc or objdump.
> ```
When does this exception get thrown? It is working fine on anrdoid. also not sure how to go about this.

**Comment by xeffyr** [comment]
Looks like it wants to use hardlinks:
```
Local Exception
Traceback (most recent call last):
  File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/archiver.py", line 4501, in main
    exit_code = archiver.run(args)
  File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/archiver.py", line 4433, in run
    return set_ec(func(args))
  File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/archiver.py", line 166, in wrapper
    return method(self, args, repository=repository, **kwargs)
  File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/archiver.py", line 280, in do_init
    key = key_creator(repository, args)
  File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/crypto/key.py", line 110, in key_creator
    return key.create(repository, args)
  File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/crypto/key.py", line 690, in create
    key.save(target, passphrase)
  File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/crypto/key.py", line 800, in save
    target.save_key(key_data)
  File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/repository.py", line 310, in save_key
    self.save_config(self.path, self.config)
  File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/repository.py", line 292, in save_config
    os.link(config_path, old_config_path)
AttributeError: module 'os' has no attribute 'link'

Platform: Linux localhost 3.18.91-15877854-QB25123770 #1 SMP PREEMPT Mon Jul 29 11:47:47 KST 2019 aarch64
Linux: Unknown Linux  
Borg: 1.1.10  Python: CPython 3.8.0 msgpack: 0.5.6
PID: 839  CWD: /data/data/com.termux/files/home
sys.argv: ['/data/data/com.termux/files/usr/bin/borg', 'init', '--encryption=repokey', './test']
SSH_ORIGINAL_COMMAND: None
```

Our python build does not support `link()` because hardlinks are disallowed by SELinux policy on Android.

**Comment by amaramrahul** [comment]
> Looks like it wants to use hardlinks:
> 
> ```
> Local Exception
> Traceback (most recent call last):
>   File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/archiver.py", line 4501, in main
>     exit_code = archiver.run(args)
>   File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/archiver.py", line 4433, in run
>     return set_ec(func(args))
>   File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/archiver.py", line 166, in wrapper
>     return method(self, args, repository=repository, **kwargs)
>   File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/archiver.py", line 280, in do_init
>     key = key_creator(repository, args)
>   File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/crypto/key.py", line 110, in key_creator
>     return key.create(repository, args)
>   File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/crypto/key.py", line 690, in create
>     key.save(target, passphrase)
>   File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/crypto/key.py", line 800, in save
>     target.save_key(key_data)
>   File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/repository.py", line 310, in save_key
>     self.save_config(self.path, self.config)
>   File "/data/data/com.termux/files/usr/lib/python3.8/site-packages/borgbackup-1.1.10-py3.8-linux-x86_64.egg/borg/repository.py", line 292, in save_config
>     os.link(config_path, old_config_path)
> AttributeError: module 'os' has no attribute 'link'
> 
> Platform: Linux localhost 3.18.91-15877854-QB25123770 #1 SMP PREEMPT Mon Jul 29 11:47:47 KST 2019 aarch64
> Linux: Unknown Linux  
> Borg: 1.1.10  Python: CPython 3.8.0 msgpack: 0.5.6
> PID: 839  CWD: /data/data/com.termux/files/home
> sys.argv: ['/data/data/com.termux/files/usr/bin/borg', 'init', '--encryption=repokey', './test']
> SSH_ORIGINAL_COMMAND: None
> ```
> 
> Our python build does not support `link()` because hardlinks are disallowed by SELinux policy on Android.
Hard links are only needed if you are running borg as a server. I have been using it for running on the client to backup my mobile to another server running borg and it has been working successfully.

**Comment by xeffyr** [comment]
`os.link` problem for `borg init` is now fixed and cross-compiled package seems to work fine even if `*.so` extensions have wrong name.

Merging.

**Pull request merged**